### PR TITLE
Implement KDBX 4.1

### PR DESCRIPTION
**Pull request by phoerious**
Implements KDBX 4.1 as described in https://keepass.info/help/kb/kdbx_4.1.html and adds test cases.

Progress:
- [x] Group Tags
- [x] Optional Password Quality Estimation
- [x] Previous Parent Groups (including simple "Restore Entry" feature)
- [x] Custom Icons with Names and Last Modification/Deletion Times
- [x] Custom Data Items with Last Modification Times

Individual commits pass tests and don't need to be squashed.

## Type of change
[NOTE]: # ( Please remove all lines which don't apply. )
- ✅ New feature (change that adds functionality)

### CODE REVIEW

**Comment by droidmonkey** [comment]
Just thinking we need to do this! We'll do this as a merge commit when the time comes.

**Comment by phoerious** [comment]
Can't do 2.7 without.

**Comment by louib** [comment]
@phoerious please keep in mind that we will remove `QImage` from `core/` in https://github.com/keepassxreboot/keepassxc/pull/6280 :pray:

**Comment by droidmonkey** [comment]
We should merge that first

File: src/core/CustomData.cpp [hunk 1]
```diff
@@ -43,7 +46,17 @@ bool CustomData::hasKey(const QString& key) const
 
 QString CustomData::value(const QString& key) const
 {
-    return m_data.value(key);
+    return m_data.value(key).value;
+}
+
+const CustomData::CustomDataItem& CustomData::item(const QString& key) const
+{
+    auto icon = m_data.find(key);
```
Comment (droidmonkey) [inline src/core/CustomData.cpp]: Why is this named icon?

File: src/core/CustomData.cpp [hunk 2]
```diff
@@ -53,20 +66,28 @@ bool CustomData::contains(const QString& key) const
...
     bool addAttribute = !m_data.contains(key);
-    bool changeValue = !addAttribute && (m_data.value(key) != value);
+    bool changeValue = !addAttribute && (m_data.value(key).value != item.value);
 
     if (addAttribute) {
         emit aboutToBeAdded(key);
     }
 
+    if (!item.lastModified.isValid()) {
+        Clock::currentDateTimeUtc();
```
Comment (droidmonkey) [inline src/core/CustomData.cpp]: Looks like you are missing an assignment here.

File: src/core/CustomData.cpp [hunk 3]
```diff
@@ -119,21 +146,52 @@ void CustomData::copyDataFrom(const CustomData* other)
     emit aboutToBeReset();
 
     m_data = other->m_data;
+    auto i = m_data.begin();
+    while (i != m_data.end()) {
+        i.value().lastModified = Clock::currentDateTimeUtc();
+        ++i;
+    }
```
Comment (droidmonkey) [inline src/core/CustomData.cpp]: I don't think it is appropriate to update the modified time when copying data. Also `updateLastModified()` is called right after this which should be doing the same thing.

File: src/core/Entry.cpp [hunk 4]
```diff
@@ -1373,7 +1378,38 @@ QString Entry::resolveUrl(const QString& url) const
...
+    return database()->rootGroup()->findGroupByUuid(m_data.previousParentGroupUuid);
+}
+
+const Group* Entry::previousParentGroup() const
+{
+    if (!database() || !database()->rootGroup()) {
+        return nullptr;
+    }
+    return database()->rootGroup()->findGroupByUuid(m_data.previousParentGroupUuid);
+}
```
Comment (droidmonkey) [inline src/core/Entry.cpp:1398]: Can we just const cast the call from the non const function? I hate duplicating code just for constness.

File: src/format/KdbxXmlReader.cpp [hunk 5]
```diff
@@ -747,6 +767,17 @@ Entry* KdbxXmlReader::parseEntry(bool history)
...
+            // Upgrade pre-KDBX-4.1 password report exclude flag
+            if (entry->customData()->contains(CustomData::ExcludeFromReportsLegacy)) {
+                entry->setExcludeFromReports(entry->customData()->value(CustomData::ExcludeFromReportsLegacy)
+                                             == TRUE_STR);
+                entry->customData()->remove(CustomData::ExcludeFromReportsLegacy);
+            }
+            continue;
+        }
+        if (m_xml.name() == "PreviousParentGroup") {
+            entry->setPreviousParentGroupUuid(readUuid());
```
Comment (droidmonkey) [inline src/format/KdbxXmlReader.cpp:780]: I'm very conflicted about burying this implementation detail into the kdbx reader. This should almost be like a post-read groom/update process.

File: src/format/KdbxXmlWriter.cpp [hunk 6]
```diff
@@ -162,12 +161,20 @@ void KdbxXmlWriter::writeCustomIcons()
...
 }
 
-void KdbxXmlWriter::writeIcon(const QUuid& uuid, const QByteArray& iconData)
+void KdbxXmlWriter::writeIcon(const QUuid& uuid, const Metadata::CustomIconData& iconData)
 {
     m_xml.writeStartElement("Icon");
 
     writeUuid("UUID", uuid);
-    writeBinary("Data", iconData);
+    if (m_kdbxVersion >= KeePass2::FILE_VERSION_4) {
```
Comment (droidmonkey) [inline src/format/KdbxXmlWriter.cpp]: Shouldn't this be > not >=  ?

File: src/format/KdbxXmlWriter.cpp [hunk 7]
```diff
@@ -222,18 +229,23 @@ void KdbxXmlWriter::writeCustomData(const CustomData* customData)
...
+void KdbxXmlWriter::writeCustomDataItem(const QString& key,
+                                        const CustomData::CustomDataItem& item,
+                                        bool writeLastModified)
 {
     m_xml.writeStartElement("Item");
 
     writeString("Key", key);
-    writeString("Value", value);
+    writeString("Value", item.value);
+    if (writeLastModified && m_kdbxVersion >= KeePass2::FILE_VERSION_4 && item.lastModified.isValid()) {
```
Comment (droidmonkey) [inline src/format/KdbxXmlWriter.cpp]: Same here, this is a kdbx 4.1 feature, not 4

File: src/format/KdbxXmlWriter.cpp [hunk 8]
```diff
@@ -345,6 +366,15 @@ void KdbxXmlWriter::writeEntry(const Entry* entry)
     writeString("Tags", entry->tags());
     writeTimes(entry->timeInfo());
 
+    if (m_kdbxVersion >= KeePass2::FILE_VERSION_4) {
```
Comment (droidmonkey) [inline src/format/KdbxXmlWriter.cpp]: Same here too

**Review by droidmonkey — CHANGES_REQUESTED** [review]

File: src/core/CustomData.cpp (hunk 1 above)
Comment (phoerious) [reply src/core/CustomData.cpp]: Good point. Probably happened during refactoring.

File: src/core/CustomData.cpp (hunk 3 above)
Comment (phoerious) [reply src/core/CustomData.cpp]: OK. I wasnt sure about what would be the correct and expected behaviour here.

File: src/core/Entry.cpp (hunk 4 above)
Comment (phoerious) [reply src/core/Entry.cpp:1398]: No, these are two different code paths, because all the find functions have a const and a non-const version (which we need both) and the KDBX writer is all const, so I couldn't call any of the non-const members from there. Most of these functions have two implementations just for this. I tried unwinding the const path to avoid duplication, but ran against a wall.

File: src/format/KdbxXmlReader.cpp (hunk 5 above)
Comment (phoerious) [reply src/format/KdbxXmlReader.cpp:780]: I think it's fine. It's basically just reading both two different things into one format.

File: src/format/KdbxXmlWriter.cpp (hunk 6 above)
Comment (phoerious) [reply src/format/KdbxXmlWriter.cpp]: No, we have no KDBX4.1 flag. Effectively, the version hasn't changed and the binary KDBX envelope has the same version as well (unless Dominik decides to change that once all alt clients support the new features). The only way to differentiate 4.0 and 4.1 is by checking if the DB contains the new features, which is useless.

Comment (phoerious) [reply src/format/KdbxXmlWriter.cpp]: Although, now that you say it, I will have to recheck if we aren't just masking out the bits on the envelope. If that is the case, than we have to stop doing that. In either case I tested the changes and they work in both KeePassXC and KeePass.

File: src/core/CustomData.cpp (hunk 3 above)
Comment (droidmonkey) [reply src/core/CustomData.cpp]: I think we were doing it wrong before by updating the last modified time attribute. Never really noticed that behavior before.

File: src/format/KdbxXmlWriter.cpp (hunk 6 above)
Comment (phoerious) [reply src/format/KdbxXmlWriter.cpp]: That was actually it. In fact, our entire loading and saving logic was a binary "KDBX4 yes or no" thing, throwing away all minor version bits.

I changed it in a way that you get a warning when trying to open a file with a newer minor version (still an error for unsupported major versions) and made KeePassXC determine the lowest possible KDBX version for saving changes dynamically. This of course requires us to save the format version in the Database object, because we never want to downgrade (we were throwing that information away previously).

File: src/core/CustomData.cpp [hunk 9]
```diff
@@ -53,20 +66,28 @@ bool CustomData::contains(const QString& key) const
...
     bool addAttribute = !m_data.contains(key);
-    bool changeValue = !addAttribute && (m_data.value(key) != value);
+    bool changeValue = !addAttribute && (m_data.value(key).value != item.value);
 
     if (addAttribute) {
         emit aboutToBeAdded(key);
     }
 
+    if (!item.lastModified.isValid()) {
+        item.lastModified = Clock::currentDateTimeUtc();
```
Comment (droidmonkey) [inline src/core/CustomData.cpp:87]: Shouldn't we set this to the current time no matter what, especially if it is valid? We do that in the `rename` function.

File: src/core/CustomData.cpp [hunk 10]
```diff
@@ -125,15 +152,41 @@ void CustomData::copyDataFrom(const CustomData* other)
...
+    if (m_data.isEmpty() || (m_data.size() == 1 && m_data.contains(LastModified))) {
+        m_data.remove(LastModified);
+        return;
+    }
+
+    if (!lastModified.isValid()) {
+        lastModified = Clock::currentDateTimeUtc();
+    }
+    m_data.insert(LastModified, {lastModified.toString(), QDateTime()});
+}
```
Comment (droidmonkey) [inline src/core/CustomData.cpp:187]: Do we want to keep doing this now that each item has a modified timestamp? My thought would be that this would be replaced on load with the modified timestamps:
1. Database read from file into memory
2. Database custom data checks for _LAST_MODIFIED entry
3. If present, set all last modified attributes of other custom data elements and delete _LAST_MODIFIED
4. On Save, this new state is set (no _LAST_MODIFIED) and the database is upgraded to KDBX 4.1

We shouldn't have _LAST_MODIFIED items in entry and group custom data since the date is derived from the entry/group modified date.

File: src/format/Kdbx3Writer.cpp [hunk 11]
```diff
@@ -137,7 +137,7 @@ bool Kdbx3Writer::writeDatabase(QIODevice* device, Database* db)
         return false;
     }
 
-    KdbxXmlWriter xmlWriter(formatVersion());
+    KdbxXmlWriter xmlWriter(db->formatVersion());
```
Comment (droidmonkey) [inline src/format/Kdbx3Writer.cpp:140]: I guess this means in theory we could have KDBX 3.1 headers and KDBX 4.1 XML insides? Would KeePass2 barf on this combination? Would we barf on it on read?

File: src/core/Database.cpp [hunk 12]
```diff
@@ -116,14 +116,15 @@ bool Database::open(QSharedPointer<const CompositeKey> key, QString* error, bool
  * @param error error message in case of failure
  * @return true on success
  */
-bool Database::open(const QString& filePath, QSharedPointer<const CompositeKey> key, QString* error, bool readOnly)
+KeePass2Reader::Status
+Database::open(const QString& filePath, QSharedPointer<const CompositeKey> key, QString* error, bool readOnly)
```
Comment (droidmonkey) [inline src/core/Database.cpp]: I really don't like the Database class returning the status type of a completely different class. That is the point of the error string to capture the actual error and to either succeed in opening (true) or fail (false). Having more than those states complicates handling of hard/soft failures.

File: src/format/KeePass2Writer.h [hunk 13]
```diff
@@ -33,6 +33,7 @@ class KeePass2Writer
     bool writeDatabase(const QString& filename, Database* db);
     bool writeDatabase(QIODevice* device, Database* db);
     void extractDatabase(Database* db, QByteArray& xmlOutput);
+    static quint32 needsKdbxVersion(Database const* db, bool ignoreCurrent = false, bool ignoreKdf = false);
```
Comment (droidmonkey) [inline src/format/KeePass2Writer.h]: I recommend renaming this to `kdbxVersionRequired` or `kdbxVersionNeeded`

File: src/gui/DatabaseOpenWidget.cpp [hunk 14]
```diff
@@ -202,11 +202,31 @@ void DatabaseOpenWidget::openDatabase()
...
-    if (ok) {
+    if (status == KeePass2Reader::Status::VersionWarn) {
+        QScopedPointer<QMessageBox> msgBox(new QMessageBox(this));
+        msgBox->setIcon(QMessageBox::Warning);
+        msgBox->setWindowTitle(tr("Database Version Mismatch"));
+        msgBox->setText(tr("The database you are trying to open was most likely\n"
+                           "created by a newer version of KeePassXC.\n\n"
+                           "You can try to open it anyway, but it may be incomplete\n"
+                           "and saving any changes may incur data loss.\n\n"
+                           "We recommend you update your KeePassXC installation."));
```
Comment (droidmonkey) [inline src/gui/DatabaseOpenWidget.cpp:217]: I don't really understand this warning, when would a user ever see this? Why not just show a warning on open that the database may be incomplete due to unsupported KDBX version? What should the user do if they are not comfortable opening the database?

**Review by droidmonkey — CHANGES_REQUESTED** [review]

File: src/core/CustomData.cpp (hunk 9 above)
Comment (phoerious) [reply src/core/CustomData.cpp:87]: No, this will only fill in a date when no date has been provided by the user of the function. We don't want to overwrite dates when one is present in the file
That would make this field useless, because it would be like changing the modification time of a file on read.

File: src/core/CustomData.cpp (hunk 10 above)
Comment (phoerious) [reply src/core/CustomData.cpp:187]: I described that in the commit message. I left the old behaviour untouched, because otherwise CustomData became unmergeable due to the lack of DeletedObjects. This thing here is ugly as hell, but I don't see a good way to avoid it right now. Having dates on CustomData is great, but useless as of now, since you cannot detect deletions. My implementation effectively only prevents that KeePassXC discards dates set in the file, but doesn't really make use of them so far.

File: src/format/Kdbx3Writer.cpp (hunk 11 above)
Comment (phoerious) [reply src/format/Kdbx3Writer.cpp:140]: We use the same XML reader with a few version checks anyway. This has always been the case and would in theory be possible. I don't know about KeePass, but KeePassXC would be able to handle that case most likely. Reading a KDBX4.1 XML with a KDBX 4 header is definitely possible. If you try that across major versions, there might be errors when reading dates or other things that are hard incompatible (maybe, never tested it). But overall, we are pretty lenient and ignore stuff we don't know or wouldn't expect in that version anyway. The envelope version bytes are a mere hint.

File: src/core/Database.cpp (hunk 12 above)
Comment (phoerious) [reply src/core/Database.cpp]: My thought was that it would make the API more transparent and explicit, because you know that this is a case you may want to handle. But you don't have to, the enum can be cast implicitly to bool if you are interested only in hard failures. This status isn't needed, I could read it from the Database object, but it would make an edge case more obvious.

I am unsure about whether we want this or not. It may make things more complicated, but it would definitely make them more explicit. Also returning an error code in combination with an error description is generally a useful thing to do (most C/C++ libraries do exactly that), because you cannot handle different errors or warnings appropriately if there's only a description in natural language.

If you are concerned only about the "other class" thing I could duplicate the enum to this class.

File: src/gui/DatabaseOpenWidget.cpp (hunk 14 above)
Comment (phoerious) [reply src/gui/DatabaseOpenWidget.cpp:217]: This can be rather critical, so I would prefer if the user has to accept the case explicitly to be 100% sure they're aware of the (or some) danger. A simple warning at the top can be dismissed too easily and things could happen that trigger a save even if the user doesn't want to. KeePassXC ignores unknown fields on read and hence discards them on write, so you can lose data.

If you want a simple banner warning, we have to bring back the read-only mode.

Besides, any useful message would be too long for a banner. This is the shortest I could make it and it's still not banner-short.

Comment (michaelk83) [reply src/gui/DatabaseOpenWidget.cpp:217]: A shorter phrasing may be:
> The database version is not fully supported. Some data may not be read, and saving changes may incur data loss. Proceed anyway?

You could add the part about upgrading, but a newer minor DB version may not come from KPXC, and a newer KPXC version may not actually be available yet. The safest approach is to refuse to open it at all.

Comment (phoerious) [reply src/gui/DatabaseOpenWidget.cpp:217]: Way too technical and way too vague. Users will be completely lost when they read that. "KeePassXC version too old" ist much clearer than "unsupported somehow for some reason". And "unsupported" is also generally a term used by tech people that scares non-technical folks.

Comment (michaelk83) [reply src/gui/DatabaseOpenWidget.cpp:217]: In this case, scaring them is probably a good thing. But you can adjust it as you see fit.

File: src/core/CustomData.cpp (hunk 9 above)
Comment (droidmonkey) [reply src/core/CustomData.cpp:87]: Sorry I missed that item was passed in to this function, not a simple value.

File: src/core/Database.cpp (hunk 12 above)
Comment (droidmonkey) [reply src/core/Database.cpp]: > This status isn't needed, I could read it from the Database object

I was thinking that as well, keep the reader simple and out of these concerns. 

Its nice to detect and report on an edge case situation. We just need to make sure that 1) the user even cares, 2) we might just want to silently fix the problem or issue a warning message, and/or 3) just hard fail when a version mismatch exists (I don't recommend).

File: src/gui/DatabaseOpenWidget.cpp (hunk 14 above)
Comment (droidmonkey) [reply src/gui/DatabaseOpenWidget.cpp:217]: Maybe this:

> You are trying to open a password database with a newer format than this version of KeePassXC supports. Continuing may result in loss of data or corruption. Would you like to continue?

Comment (phoerious) [reply src/gui/DatabaseOpenWidget.cpp:217]: I don't like the word format there. It's too technical again. I tried to dumb it down as much as possible. Technical people will understand it anyway and other users are not thrown off entirely.

File: src/core/Database.cpp (hunk 12 above)
Comment (phoerious) [reply src/core/Database.cpp]: I don't really care about the user of the application here. I'm thinking more of users of our API, who are forced to think about why there's an enum returned.

Comment (phoerious) [reply src/core/Database.cpp]: I moved the `hasMinorVersionMismatch()` method into the `Database` class and reverted it back to a bool. That's still not as explicit, but more explicit than completely ignoring the fact that there might be a minor version mismatch.

Comment (droidmonkey) [reply src/core/Database.cpp]: Part of me wants to make a DatabaseFileHandler class eventually and move all the open/save into it. Technically the KeePass2Reader should return a QSharedPointer&lt;Database&gt; that is subsequently used instead of writing into a provided Database instance. Almost all classes would interact with the DatabaseFileHandler class.

File: src/gui/DatabaseOpenWidget.cpp (hunk 14 above)
Comment (michaelk83) [reply src/gui/DatabaseOpenWidget.cpp:217]: IMO, if continuing may result in data loss, it's best not to give that option at all (unless really necessary, like with the different save modes). Users may just as easily click through a prompt without paying enough attention, and then wonder what happened to their data.

If you can detect all unsupported features when loading, then you can hard-fail only when such features are detected. Otherwise, why risk it? Or you can bring back read-only mode, as suggested.

File: src/core/Database.cpp (hunk 12 above)
Comment (phoerious) [reply src/core/Database.cpp]: TBH, I would rather have a Database class with a load() method that manages the allocated resources RAII-style and then have it call a bunch of free functions for doing the actual parsing and loading stuff. Just because it's Qt, doesn't mean we have to do everything OOP (thank goodness).

**Review by droidmonkey — APPROVED** [review]
